File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public boolean findOne = false;
    public boolean findTwo = false;
    public TreeNode sa;
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (p.val > q.val)
        {
            TreeNode temp = p;
            p = q;
            q = temp;
        }
        findLow(root, p, q);
        return sa;
    }
    
    public void findLow(TreeNode root, TreeNode p, TreeNode q)
    {
        if (root != null && sa == null)
        {
            if (root.val >= p.val && root.val <= q.val)
            {
                sa = root;
            }
            if (root.val > p.val && root.val > q.val)
            {
                findLow(root.left, p, q);
            }
            else
            {
                findLow(root.right, p, q);
            }
        }
    }
        
}

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        while root:
            if root.val > p.val and root.val > q.val:
                root = root.left
            elif root.val < p.val and root.val < q.val:
                root = root.right
            else:
                return root

        # if root == None:
        #     return None
        # def dfs(node):
        #     if node == None:
        #         return None
        #     if node == p or node == q:
        #         return node
        #     left = dfs(node.left)
        #     right = dfs(node.right)
        #     if left and right:
        #         return node
        #     return left if left else right
        # return dfs(root)    